### DISCUSSION TITLE
Converters to and from

### DISCUSSION THREAD

**Discussion by sonicsmooth** [AvaloniaUI/Avalonia#10146] [General]
I have a few questions about converters, in this case user string --> value --> formatted string
I have a user-typed text field which should be converted from a string with possible engineering notation, to an internal numerical value.
I also need to convert that internal numerical value to a properly formatted string in the UI TextBlock.

          <TextBox Text="{TemplateBinding DValue,
                          Converter={StaticResource myDistanceToStringConverter},
                          Mode=OneWayToSource,
                          ConverterParameter=um}" />

1. I'm not sure which is the "normal" direction for the converter.  I think the "normal' direction (`IValueConverter.Convert`) is from Avalonia Property to UI, and that `IValueConverter.ConvertBack` is the opposite, but I'm not sure whether `ConvertBack` is used only in error cases or normal.  It seemed to be called at startup when the user textbox was being automatically populated, and this happened even with `OneWayToSource`.
2. I only need the direction to be from the UI to the property in this case.  Changing `Mode=OneWayToSource` or `TwoWay` didn't help matters.
3. What should I return when the value cannot be converted, such as if the user types "hello" instead of a proper number such as "15 mm".  I tried returning `null`, `new BindingNotification(new InvalidCastException(), BindingErrorType.Error)`, and throwing `new NotSupportedException()`, but these did not result in the red mark showing up in the UI. 
4. The underlying property is a custom type `Distance`, with a question mark `Distance?` so it can be null if the conversion fails, but that didn't work.
5. Should I somehow try validating this?  Would this validation occur before the conversion is attempted?
6. Given https://groups.google.com/g/wpf-disciples/c/P-JwzRB_GE8 should I just do the conversions "manually" in the setter and handling events?  I can instantiate three properties -- UserString, DValue, OutputString.  It's not as elegant I suppose.
7. This is in a `TemplatedControl` and all the properties are defined in codebehind, not viewmodel, if that makes any difference.

**Comment by timunie** [comment]
In a TemplatedControl, I probably would calculate the other value in code behind. You can check out `ColorPicker` or other controls for example: https://github.com/AvaloniaUI/Avalonia/blob/5896e3328676ec0aaf36b35f09356051f88c24c6/src/Avalonia.Controls.ColorPicker/ColorSlider/ColorSlider.cs#L173

**Reply by sonicsmooth** [reply]
thanks will check it out.  I've been meaning to user the new color picker too...
